### PR TITLE
Merge plugins into the HLS package

### PR DESCRIPTION
**Pull request by michaelpj**
This is a POC for merging all the plugin packages into the HLS package, as discussed in https://github.com/haskell/haskell-language-server/issues/3556.

I just did the first two plugins alphabetically to show what it looks like, it's pretty mechanical.

To reiterate, the advantages are:
- Fewer packages to upload
- Less repeated metadata
- Fewer versions
- Ability to use common stanzas to e.g. have the same set of warnings everywhere (I would do this as a follow up if we decide to go this route)

Things seem to still work fine. The main annoyances are:
- Can't `cd` into a plugin directory and `cabal build`/`cabal test`
- Can't `cabal test <plugin-name>`, instead need `cabal test <plugin-name>-tests`

I think overall it's still going to be a pretty big improvement

### CODE REVIEW

**Comment by soulomoon** [comment]
One high draw back, it prevent us from working on a single plugin along.

**Comment by michaelpj** [comment]
I don't think it does? You can still just build the component for the plugin and it won't build anything else.

**Comment by soulomoon** [comment]
But we can no longer open an single editor containing just one plugin dir and get hls support 🤔

**Comment by michaelpj** [comment]
I don't see why not? We do support single components of packages just fine? And it'll be better as the multi-component stuff rolls out (but that's already bad if you try to edit multiple packages today)

**Comment by soulomoon** [comment]
> I don't see why not? We do support single components of packages just fine? And it'll be better as the multi-component stuff rolls out (but that's already bad if you try to edit multiple packages today)

Thanx for clearing out my concern on this.

**Comment by michaelpj** [comment]
Another downside: stack supports building packages with internal libraries (the CI run succeeded!), but it doesn't do per-component builds, so the dev experience using stack will probably suck. I'm not sure how much we care about that these days :shrug:

**Comment by michaelpj** [comment]
No longer a POC! 

There is still plenty to do here with cleaning up the now rather monolithic cabal file, making better use of common stanzas etc, but I'm keen to land this before too long since it's very conflict-y.

**Comment by michaelpj** [comment]
cc @fendor @wz1000  as discussed

**Comment by michaelpj** [comment]
Follow up todos:
- Make all components which condition on `pedantic` use the common stanza for that
- Add a common stanza for `default-language`
- Maybe merge in some more packages. I think only `hls-test-utils`, everything else is a dependency of `ghcide` and so needs to stay out for now.
- Maybe try and organize the cabal file to be a bit more navigable

File: haskell-language-server.cabal [hunk 1]
```diff
@@ -63,286 +69,1679 @@ flag ignore-plugins-ghc-bounds
...
+    , text
+    , transformers
+    , unordered-containers
+
+  default-language:   Haskell2010
+  default-extensions:
+    DataKinds
+    TypeOperators
+
+  ghc-options:        -Wno-unticked-promoted-constructors
```
Comment (jhrcek) [inline haskell-language-server.cabal:588]: Can be removed - is now in common warnings.

File: haskell-language-server.cabal [hunk 2]
```diff
@@ -63,286 +69,1679 @@ flag ignore-plugins-ghc-bounds
...
+    , temporary
+    , text
+    , transformers
+    , unordered-containers
+    , ghc-lib-parser
+    , ghc-lib-parser-ex
+    , apply-refact
+
+  cpp-options:   -DHLINT_ON_GHC_LIB
+  ghc-options:
```
Comment (jhrcek) [inline haskell-language-server.cabal:656]: You can import common warnings for this lib and remove this.

**Review by jhrcek — COMMENTED** [review]
I like this approach more than the previous.
Few nitpicks about ghc-options - warnings, but I will work on those anyway in followup PRs, so feel free to ignore.

File: haskell-language-server.cabal [hunk 3]
```diff
@@ -63,286 +69,1679 @@ flag ignore-plugins-ghc-bounds
...
+  default-language: Haskell2010
+
+test-suite hls-refactor-plugin-tests
+  import: warnings
+  type:             exitcode-stdio-1.0
+  default-language: Haskell2010
+  hs-source-dirs:   plugins/hls-refactor-plugin/test
+  main-is:          Main.hs
+  other-modules:    Test.AddArgument
+  ghc-options:      -O0 -threaded -rtsopts -with-rtsopts=-N -Wunused-imports
```
Comment (jhrcek) [inline haskell-language-server.cabal:1633]: I'm pretty sure that -Wunused-imports are enabled by -Wall, so you could remove it from here since you're importing -Wall via `import: warnings`

**Comment by michaelpj** [comment]
Yeah, I didn't do any cleanup on the cabal file contents, just moved them over as-is. Cleanups later!

**Comment by michaelpj** [comment]
Hmmm... is that a long-path error?

**Comment by michaelpj** [comment]
Yup, the path is 261 characters. I'm going to shorten the component name. Generally I think we could now strip the `hls-` prefixes from everything at least. Frankly, maybe also the `-plugin` suffix?

**Comment by michaelpj** [comment]
Still can't get everything below the limit. At this point it's either do something ad-hoc or turn off GHC9.2+windows

**Comment by fendor** [comment]
> Generally I think we could now strip the hls- prefixes from everything at least. Frankly, maybe also the -plugin suffix?

Changing `hls-fourmolu-plugin` to `fourmolu` will cause a fair amount of confusion. Almost all plugins that essentially wrap a tool would be weird that way.

**Comment by michaelpj** [comment]
Yeah, I think we should probably keep `-plugin`.

**Comment by fendor** [comment]
Personally, I like the explicitness :shrug: It signals an HLS namespace

**Comment by michaelpj** [comment]
Right, but once it's a component of the `haskell-language-server` package it's already namespaced. The full name will be `haskell-language-server:hls-ormolu-plugin` which seems redundant to me.

**Comment by michaelpj** [comment]
You definitely want it when it's _not_ a subcomponent.

**Comment by michaelpj** [comment]
Okay, I think this is now all working, just getting through CI nondeterminism. I've reverted all component name changes for now since I've sidestepped the long path issue. We can discuss that later.

Anyone want to approve?

**Review by joyfulmantis — APPROVED** [review]
LGTM

**Review by fendor — CHANGES_REQUESTED** [review]
I think the tests are not run in CI 
image

**Comment by georgefst** [comment]
I suppose the `fourmolu`/`fourmolu-plugin`/`hls-fourmolu-plugin` trade-off comes down to whether sublibraries can only over be referred to outside the package in qualified form. I've dabbled a little with multiple-public-libraries support, but I don't know this off the top of my head, though I'd assume they can't.

In general, this change seems a sensible one. I had assumed when the project started back in 2020 that we'd eventually move towards plugins being maintained outside of the HLS organisation entirely, but this doesn't really work without them being genuinely runtime-pluginnable. And as far as I understand it, this would require significant improvements in GHC dynamic linking support, which no one wants to work on...

tl;dr All good.

**Comment by joyfulmantis** [comment]
> I think the tests are not run in CI image

Yikes! yes we definitely want those tests still

**Comment by michaelpj** [comment]
> I think the tests are not run in CI

What on earth... that is so odd. It works fine for me locally. Investigating

**Comment by michaelpj** [comment]
> I suppose the fourmolu/fourmolu-plugin/hls-fourmolu-plugin trade-off comes down to whether sublibraries can only over be referred to outside the package in qualified form

Note that these are not currently _public_ sublibraries. We could make them public, but I don't know that there are any actual consumers of them. So at the moment we're only using the much older and better-supported non-public internal libraries.

> I had assumed when the project started back in 2020 that we'd eventually move towards plugins being maintained outside of the HLS organisation entirely, but this doesn't really work without them being genuinely runtime-pluginnable

Yes, I think the _reality_ of how plugins are used by HLS is that they are a very handy method of code organization and responsibility allocation, but that's about it. And I think that's fine! My position has always been that nobody actually wants to do runtime customization of their HLS installation, they want it to just work.

**Comment by michaelpj** [comment]
My suspicion is that the test issue was due to tasty-rerun actually working. I've deleted the caches relating to that from this PR and rerun everything, let's see if it now runs the tests.

**Comment by georgefst** [comment]
> Note that these are not currently _public_ sublibraries.

Oh, I just assumed. It is _in principle_ a regression to make all these libraries unusable externally. Though I suspect no one's using them.

**Comment by michaelpj** [comment]
Yes, it's for sure a regression in theory. But personally I _highly_ doubt that anyone is using them. You'd have to be assembling your own custom HLS, and not want to do it using HLS's cabal flags for some reason.

**Comment by michaelpj** [comment]
Nope?

**Comment by michaelpj** [comment]
It was tasty-rerun, turning it off makes the tests run. I don't know whether it's doing something dodgy, but it seems to be orthogonal to this PR at least.

**Comment by smunix** [comment]
@michaelpj - it looks like this change broke `nix` rpath setting for plugins internal libraries. It fails during the `install_phase` for finding references to the `/build/` directory. Have `2.7.0.0` been installed successfully with `nix` at all?

**Comment by michaelpj** [comment]
That seems like a question for the Nix Haskell folks? It builds with cabal, so if it doesn't build with the Nix setup then that's a bug in the Nix setup, I think.
